File: rust/chatgear/src/command.rs
//! Device command types.

use giztoy_jsontime::Milli;
use serde::{Deserialize, Serialize};
use std::fmt;

/// The interface for device commands.
pub trait SessionCommand: fmt::Debug + Send + Sync {
    /// Returns the command type string.
    fn command_type(&self) -> &'static str;
}

/// Wraps a command with metadata.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionCommandEvent {
    /// Command type identifier.
    #[serde(rename = "type")]
    pub cmd_type: String,

    /// Event timestamp.
    pub time: Milli,

    /// Command payload (serialized).
    #[serde(rename = "pld")]
    pub payload: serde_json::Value,

    /// When the command was issued.
    pub issue_at: Milli,
}

impl SessionCommandEvent {
    /// Creates a new command event.
    pub fn new<C: SessionCommand + Serialize>(cmd: &C) -> Self {
        let now = Milli::now();
        Self {
            cmd_type: cmd.command_type().to_string(),
            time: now,
            payload: serde_json::to_value(cmd).unwrap_or(serde_json::Value::Null),
            issue_at: now,
        }
    }

    /// Creates a new command event with a specific issue time.
    pub fn with_issue_at<C: SessionCommand + Serialize>(cmd: &C, issue_at: Milli) -> Self {
        Self {
            cmd_type: cmd.command_type().to_string(),
            time: Milli::now(),
            payload: serde_json::to_value(cmd).unwrap_or(serde_json::Value::Null),
            issue_at,
        }
    }

    /// Attempts to deserialize the payload into a specific command type.
    pub fn parse_payload<C: for<'de> Deserialize<'de>>(&self) -> Result<C, serde_json::Error> {
        serde_json::from_value(self.payload.clone())
    }
}

// ============================================================================
// Command Types
// ============================================================================

/// Command to start/stop audio streaming.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Streaming(pub bool);

impl Streaming {
    pub fn new(enabled: bool) -> Self {
        Self(enabled)
    }
}

impl SessionCommand for Streaming {
    fn command_type(&self) -> &'static str {
        "streaming"
    }
}

/// Command to reset the device.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Reset {
    #[serde(default, skip_serializing_if = "is_false")]
    pub unpair: bool,
}

fn is_false(b: &bool) -> bool {
    !*b
}

impl Reset {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_unpair() -> Self {
        Self { unpair: true }
    }
}

impl SessionCommand for Reset {
    fn command_type(&self) -> &'static str {
        "reset"
    }
}

/// Command to raise an event (e.g., start a call).
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Raise {
    #[serde(default)]
    pub call: bool,
}

impl Raise {
    pub fn call() -> Self {
        Self { call: true }
    }
}

impl SessionCommand for Raise {
    fn command_type(&self) -> &'static str {
        "raise"
    }
}

/// Command to halt device operation.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Halt {
    #[serde(default, skip_serializing_if = "is_false")]
    pub sleep: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub shutdown: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub interrupt: bool,
}

impl Halt {
    pub fn sleep() -> Self {
        Self {
            sleep: true,
            ..Default::default()
        }
    }

    pub fn shutdown() -> Self {
        Self {
            shutdown: true,
            ..Default::default()
        }
    }

    pub fn interrupt() -> Self {
        Self {
            interrupt: true,
            ..Default::default()
        }
    }
}

impl SessionCommand for Halt {
    fn command_type(&self) -> &'static str {
        "halt"
    }
}

/// Command to set audio volume.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct SetVolume(pub i32);

impl SetVolume {
    pub fn new(volume: i32) -> Self {
        Self(volume)
    }
}

impl SessionCommand for SetVolume {
    fn command_type(&self) -> &'static str {
        "set_volume"
    }
}

/// Command to set display brightness.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct SetBrightness(pub i32);

impl SetBrightness {
    pub fn new(brightness: i32) -> Self {
        Self(brightness)
    }
}

impl SessionCommand for SetBrightness {
    fn command_type(&self) -> &'static str {
        "set_brightness"
    }
}

/// Command to set light mode.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct SetLightMode(pub String);

impl SetLightMode {
    pub fn new(mode: impl Into<String>) -> Self {
        Self(mode.into())
    }
}

impl SessionCommand for SetLightMode {
    fn command_type(&self) -> &'static str {
        "set_light_mode"
    }
}

/// Command to configure WiFi.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SetWifi {
    pub ssid: String,
    pub security: String,
    pub password: String,
}

impl SetWifi {
    pub fn new(ssid: impl Into<String>, security: impl Into<String>, password: impl Into<String>) -> Self {
        Self {
            ssid: ssid.into(),
            security: security.into(),
            password: password.into(),
        }
    }
}

impl SessionCommand for SetWifi {
    fn command_type(&self) -> &'static str {
        "set_wifi"
    }
}

/// Command to delete a stored WiFi network.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct DeleteWifi(pub String);

impl DeleteWifi {
    pub fn new(ssid: impl Into<String>) -> Self {
        Self(ssid.into())
    }
}

impl SessionCommand for DeleteWifi {
    fn command_type(&self) -> &'static str {
        "delete_wifi"
    }
}

/// OTA info for a component.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ComponentOTA {
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub version: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub image_url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub image_md5: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub data_file_url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub data_file_md5: String,
}

/// Command to initiate firmware upgrade.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct OTA {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub version: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub image_url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub image_md5: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub data_file_url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub data_file_md5: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub components: Vec<ComponentOTA>,
}

impl SessionCommand for OTA {
    fn command_type(&self) -> &'static str {
        "ota_upgrade"
    }
}

/// Enum wrapper for all command types (for deserialization).
#[derive(Debug, Clone)]
pub enum Command {
    Streaming(Streaming),
    Reset(Reset),
    Raise(Raise),
    Halt(Halt),
    SetVolume(SetVolume),
    SetBrightness(SetBrightness),
    SetLightMode(SetLightMode),
    SetWifi(SetWifi),
    DeleteWifi(DeleteWifi),
    OTA(OTA),
}

impl Command {
    /// Parses a command from a type string and JSON payload.
    pub fn parse(cmd_type: &str, payload: &serde_json::Value) -> Result<Self, serde_json::Error> {
        match cmd_type {
            "streaming" => Ok(Command::Streaming(serde_json::from_value(payload.clone())?)),
            "reset" => Ok(Command::Reset(serde_json::from_value(payload.clone())?)),
            "raise" => Ok(Command::Raise(serde_json::from_value(payload.clone())?)),
            "halt" => Ok(Command::Halt(serde_json::from_value(payload.clone())?)),
            "set_volume" => Ok(Command::SetVolume(serde_json::from_value(payload.clone())?)),
            "set_brightness" => Ok(Command::SetBrightness(serde_json::from_value(payload.clone())?)),
            "set_light_mode" => Ok(Command::SetLightMode(serde_json::from_value(payload.clone())?)),
            "set_wifi" => Ok(Command::SetWifi(serde_json::from_value(payload.clone())?)),
            "delete_wifi" => Ok(Command::DeleteWifi(serde_json::from_value(payload.clone())?)),
            "ota_upgrade" => Ok(Command::OTA(serde_json::from_value(payload.clone())?)),
            _ => Err(serde::de::Error::custom(format!("unknown command type: {}", cmd_type))),
        }
    }
}

#[cfg(test)]
mod command_tests {
    use super::*;

    #[test]
    fn test_streaming_command() {
        let cmd = Streaming::new(true);
        assert_eq!(cmd.command_type(), "streaming");

        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(json, "true");

        let restored: Streaming = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.0, true);
    }

    #[test]
    fn test_set_volume_command() {
        let cmd = SetVolume::new(75);
        assert_eq!(cmd.command_type(), "set_volume");

        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(json, "75");

        let restored: SetVolume = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.0, 75);
    }

    #[test]
    fn test_reset_command() {
        let cmd = Reset::new();
        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(json, "{}");

        let cmd = Reset::with_unpair();
        let json = serde_json::to_string(&cmd).unwrap();
        assert!(json.contains("unpair"));
    }

    #[test]
    fn test_halt_command() {
        let cmd = Halt::sleep();
        let json = serde_json::to_string(&cmd).unwrap();
        assert!(json.contains("sleep"));
        assert!(!json.contains("shutdown"));

        let cmd = Halt::shutdown();
        let json = serde_json::to_string(&cmd).unwrap();
        assert!(json.contains("shutdown"));
    }

    #[test]
    fn test_set_wifi_command() {
        let cmd = SetWifi::new("MyNetwork", "WPA2", "password123");
        assert_eq!(cmd.command_type(), "set_wifi");

        let json = serde_json::to_string(&cmd).unwrap();
        assert!(json.contains("MyNetwork"));
        assert!(json.contains("WPA2"));
        assert!(json.contains("password123"));
    }

    #[test]
    fn test_ota_command() {
        let cmd = OTA {
            version: "1.2.3".to_string(),
            image_url: "https://example.com/firmware.bin".to_string(),
            ..Default::default()
        };
        assert_eq!(cmd.command_type(), "ota_upgrade");

        let json = serde_json::to_string(&cmd).unwrap();
        assert!(json.contains("1.2.3"));
        assert!(json.contains("firmware.bin"));
    }

    #[test]
    fn test_session_command_event() {
        let cmd = SetVolume::new(50);
        let event = SessionCommandEvent::new(&cmd);

        assert_eq!(event.cmd_type, "set_volume");
        assert_eq!(event.payload, serde_json::json!(50));
    }

    #[test]
    fn test_command_parse() {
        let payload = serde_json::json!(true);
        let cmd = Command::parse("streaming", &payload).unwrap();
        match cmd {
            Command::Streaming(s) => assert_eq!(s.0, true),
            _ => panic!("expected Streaming"),
        }

        let payload = serde_json::json!({"ssid": "test", "security": "WPA2", "password": "pass"});
        let cmd = Command::parse("set_wifi", &payload).unwrap();
        match cmd {
            Command::SetWifi(w) => {
                assert_eq!(w.ssid, "test");
                assert_eq!(w.security, "WPA2");
            }
            _ => panic!("expected SetWifi"),
        }
    }
}

File: go/pkg/chatgear/command.go
package chatgear

import (
	"bytes"
	"encoding/json"
	"fmt"
	"time"

	"github.com/haivivi/giztoy/go/pkg/jsontime"
)

// Ensure all command types implement Command.
var (
	_ Command = (*Streaming)(nil)
	_ Command = (*Reset)(nil)
	_ Command = (*SetVolume)(nil)
	_ Command = (*SetBrightness)(nil)
	_ Command = (*SetLightMode)(nil)
	_ Command = (*SetWifi)(nil)
	_ Command = (*DeleteWifi)(nil)
	_ Command = (*OTA)(nil)
	_ Command = (*Raise)(nil)
	_ Command = (*Halt)(nil)
)

// Command is the interface for device commands.
type Command interface {
	isCommand()
	commandType() string
}

// CommandEvent wraps a command with metadata.
type CommandEvent struct {
	Type    string         `json:"type"`
	Time    jsontime.Milli `json:"time"`
	Payload Command        `json:"pld"`
	IssueAt jsontime.Milli `json:"issue_at"`
}

// NewCommandEvent creates a new command event.
func NewCommandEvent(cmd Command, issueAt time.Time) *CommandEvent {
	return &CommandEvent{
		Type:    cmd.commandType(),
		Time:    jsontime.NowEpochMilli(),
		Payload: cmd,
		IssueAt: jsontime.Milli(issueAt),
	}
}

// UnmarshalJSON implements json.Unmarshaler.
func (e *CommandEvent) UnmarshalJSON(b []byte) error {
	var v struct {
		Type    string          `json:"type"`
		Time    jsontime.Milli  `json:"time"`
		Payload json.RawMessage `json:"pld"`
		IssueAt jsontime.Milli  `json:"issue_at"`
	}
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	var cmd Command
	switch v.Type {
	case "streaming":
		cmd = new(Streaming)
	case "reset":
		cmd = new(Reset)
	case "set_volume":
		cmd = new(SetVolume)
	case "set_brightness":
		cmd = new(SetBrightness)
	case "set_light_mode":
		cmd = new(SetLightMode)
	case "set_wifi":
		cmd = new(SetWifi)
	case "delete_wifi":
		cmd = new(DeleteWifi)
	case "ota_upgrade":
		cmd = new(OTA)
	case "raise":
		cmd = new(Raise)
	case "halt":
		cmd = new(Halt)
	default:
		return fmt.Errorf("unknown command type: %s", v.Type)
	}

	if err := json.Unmarshal(v.Payload, cmd); err != nil {
		return err
	}

	*e = CommandEvent{
		Type:    v.Type,
		Time:    v.Time,
		Payload: cmd,
		IssueAt: v.IssueAt,
	}
	return nil
}

// Streaming is a command to start/stop audio streaming.
type Streaming bool

// NewStreaming creates a new Streaming command.
func NewStreaming(enabled bool) *Streaming {
	s := Streaming(enabled)
	return &s
}

func (*Streaming) isCommand()          {}
func (*Streaming) commandType() string { return "streaming" }

func (s Streaming) MarshalJSON() ([]byte, error) {
	return json.Marshal(bool(s))
}

func (s *Streaming) UnmarshalJSON(b []byte) error {
	var v bool
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*s = Streaming(v)
	return nil
}

// Reset is a command to reset the device.
type Reset struct {
	Unpair bool `json:"unpair,omitempty"`
}

func (*Reset) isCommand()          {}
func (*Reset) commandType() string { return "reset" }

func (r Reset) MarshalJSON() ([]byte, error) {
	if r == (Reset{}) {
		return json.Marshal(nil)
	}
	v := struct {
		Unpair bool `json:"unpair"`
	}{
		Unpair: r.Unpair,
	}
	return json.Marshal(v)
}

func (r *Reset) UnmarshalJSON(b []byte) error {
	if bytes.Equal(b, []byte("null")) {
		*r = Reset{}
		return nil
	}
	var v struct {
		Unpair bool `json:"unpair"`
	}
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*r = Reset{Unpair: v.Unpair}
	return nil
}

// Raise is a command to raise an event (e.g., start a call).
type Raise struct {
	Call bool `json:"call,omitempty"`
}

func (*Raise) isCommand()          {}
func (*Raise) commandType() string { return "raise" }

func (r Raise) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Call bool `json:"call"`
	}{
		Call: r.Call,
	})
}

func (r *Raise) UnmarshalJSON(b []byte) error {
	var v struct {
		Call bool `json:"call"`
	}
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*r = Raise{Call: v.Call}
	return nil
}

// Halt is a command to halt device operation.
type Halt struct {
	Sleep     bool `json:"sleep,omitempty"`
	Shutdown  bool `json:"shutdown,omitempty"`
	Interrupt bool `json:"interrupt,omitempty"`
}

func (*Halt) isCommand()          {}
func (*Halt) commandType() string { return "halt" }

func (h Halt) MarshalJSON() ([]byte, error) {
	if h == (Halt{}) {
		return json.Marshal(nil)
	}
	v := struct {
		Sleep     bool `json:"sleep"`
		Shutdown  bool `json:"shutdown"`
		Interrupt bool `json:"interrupt"`
	}{
		Sleep:     h.Sleep,
		Shutdown:  h.Shutdown,
		Interrupt: h.Interrupt,
	}
	return json.Marshal(v)
}

func (h *Halt) UnmarshalJSON(b []byte) error {
	if bytes.Equal(b, []byte("null")) {
		*h = Halt{}
		return nil
	}
	var v struct {
		Sleep     bool `json:"sleep"`
		Shutdown  bool `json:"shutdown"`
		Interrupt bool `json:"interrupt"`
	}
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*h = Halt{
		Sleep:     v.Sleep,
		Shutdown:  v.Shutdown,
		Interrupt: v.Interrupt,
	}
	return nil
}

// SetVolume is a command to set audio volume.
type SetVolume int

// NewSetVolume creates a new SetVolume command.
func NewSetVolume(volume int) *SetVolume {
	v := SetVolume(volume)
	return &v
}

func (*SetVolume) isCommand()          {}
func (*SetVolume) commandType() string { return "set_volume" }

func (s SetVolume) MarshalJSON() ([]byte, error) {
	return json.Marshal(int(s))
}

func (s *SetVolume) UnmarshalJSON(b []byte) error {
	var v int
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*s = SetVolume(v)
	return nil
}

// SetBrightness is a command to set display brightness.
type SetBrightness int

// NewSetBrightness creates a new SetBrightness command.
func NewSetBrightness(brightness int) *SetBrightness {
	b := SetBrightness(brightness)
	return &b
}

func (*SetBrightness) isCommand()          {}
func (*SetBrightness) commandType() string { return "set_brightness" }

func (s SetBrightness) MarshalJSON() ([]byte, error) {
	return json.Marshal(int(s))
}

func (s *SetBrightness) UnmarshalJSON(b []byte) error {
	var v int
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*s = SetBrightness(v)
	return nil
}

// SetLightMode is a command to set light mode.
type SetLightMode string

// NewSetLightMode creates a new SetLightMode command.
func NewSetLightMode(mode string) *SetLightMode {
	m := SetLightMode(mode)
	return &m
}

func (*SetLightMode) isCommand()          {}
func (*SetLightMode) commandType() string { return "set_light_mode" }

func (s SetLightMode) MarshalJSON() ([]byte, error) {
	return json.Marshal(string(s))
}

func (s *SetLightMode) UnmarshalJSON(b []byte) error {
	var v string
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*s = SetLightMode(v)
	return nil
}

// SetWifi is a command to configure WiFi.
type SetWifi struct {
	SSID     string `json:"ssid"`
	Security string `json:"security"`
	Password string `json:"password"`
}

func (SetWifi) isCommand()          {}
func (SetWifi) commandType() string { return "set_wifi" }

// DeleteWifi is a command to delete a stored WiFi network.
type DeleteWifi string

func (DeleteWifi) isCommand()          {}
func (DeleteWifi) commandType() string { return "delete_wifi" }

func (s DeleteWifi) MarshalJSON() ([]byte, error) {
	return json.Marshal(string(s))
}

func (s *DeleteWifi) UnmarshalJSON(b []byte) error {
	var v string
	if err := json.Unmarshal(b, &v); err != nil {
		return err
	}
	*s = DeleteWifi(v)
	return nil
}

// OTA is a command to initiate firmware upgrade.
type OTA struct {
	Version     string         `json:"version,omitzero"`
	ImageURL    string         `json:"image_url,omitzero"`
	ImageMD5    string         `json:"image_md5,omitzero"`
	DataFileURL string         `json:"data_file_url,omitzero"`
	DataFileMD5 string         `json:"data_file_md5,omitzero"`
	Components  []ComponentOTA `json:"components,omitzero"`
}

// ComponentOTA contains OTA info for a component.
type ComponentOTA struct {
	Name        string `json:"name"`
	Version     string `json:"version,omitzero"`
	ImageURL    string `json:"image_url,omitzero"`
	ImageMD5    string `json:"image_md5,omitzero"`
	DataFileURL string `json:"data_file_url,omitzero"`
	DataFileMD5 string `json:"data_file_md5,omitzero"`
}

func (*OTA) isCommand()          {}
func (*OTA) commandType() string { return "ota_upgrade" }
